### PR TITLE
[Feature] Support logical single table sync materialized view

### PR DESCRIPTION
**Pull request by LiShuMing**
Fixes #22401

### Backgroud
Now single table/sync mv is like an index of table which cannot be detached from the base table. This PR supports to create a sync mv  that can be a specific target olap table, so the sync mv can be a separate table from the base table. This can get much profits:
- The target mv table can have different TTLs from the base table after decoupling;
- The target mv table can be set from many base tables to achieve the final aggregation.

we can this sync mv with target table `Logical MV`.

### Implement
The logical mv is more like an ETL rather than an index, it reuses the sync mv's implements  to send the source data to both the base table and the target table in `OlapTableSink`.

NOTE:
- The logical mv does not support mv rewrite because the logical mv doesn't refresh the historic datas of the base table.

### Limits
But there are still much limits for now:
-  The expressions that the sync mv can support are still not changed because the logical mv reuse the basic sync mv's implements
- The target table of the logical mv should also have much limits:
  - the target table must have the column that stores the mv's final output;
  - the target table must have the same redistributions with the base table, or the target table have the random redistributions.
- The logical mv does not support mv rewrite because the logical mv doesn't refresh the historic datas of the base table.

### Usage

```
drop table if exists test1;
CREATE TABLE `test1` (
  `k1` tinyint(4) NULL,
  `k2` varchar(64) NULL,
  `k3` bigint NULL ,
  `k4` varchar(64) NULL
) ENGINE=OLAP 
DUPLICATE KEY(`k1`)
DISTRIBUTED BY HASH(`k1`)BUCKETS 3
PROPERTIES (
      "colocate_with" = "logical_mv"
);

drop table if exists target2;
CREATE TABLE `target2` (
  `k1` tinyint(4) NULL,
  `count_k2` bigint NULL,
  `sum_k3` bigint NULL
) ENGINE=OLAP 
DUPLICATE KEY(`k1`)
DISTRIBUTED BY HASH(`k1`)BUCKETS 3
PROPERTIES (
      "colocate_with" = "logical_mv"
);

CREATE MATERIALIZED VIEW test_mv2
to target2
as
select k1, count(k2) as count_k2, sum(k3) as sum_k3 from test1 group by k1;

insert into test1 values (4, 'a', 1, 'aa'), (5, 'b', 2, 'bb'), (6, NULL, NULL, NULL);
select * from test1 order by k1;
select * from target1 order by k1;
```

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 108 / 364 (29.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Table.java|0|2|00.00%|[230, 354]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|1|00.00%|[1126]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|0|1|00.00%|[115]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[2900, 2901]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|10|00.00%|[244, 245, 246, 247, 248, 249, 250, 251, 1163, 1164]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|4|00.00%|[253, 254, 255, 256]|
|:large_blue_circle:|com/starrocks/persist/CreateMaterializedIndexMetaInfo.java|0|34|00.00%|[30, 37, 39, 42, 43, 44, 45, 46, 47, 50, 54, 58, 62, 66, 67, 68, 69, 70, 73, 74, 75, 76, 77, 80, 81, 82, 87, 92, 93, 95, 96, 98, 99, 100]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|27|03.70%|[1640, 1641, 2088, 2089, 2090, 2092, 2093, 2094, 2096, 2097, 2098, 2100, 2102, 2103, 2104, 2106, 2107, 2108, 2109, 2110, 2111, 2112, 2113, 2114, 2116, 2119]|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|16|123|13.01%|[225, 251, 252, 253, 254, 255, 257, 258, 259, 261, 263, 264, 268, 269, 270, 271, 272, 273, 274, 275, 278, 279, 280, 281, 282, 283, 284, 285, 287, 288, 289, 294, 295, 296, 297, 298, 301, 302, 307, 308, 310, 311, 314, 315, 316, 317, 318, 320, 322, 323, 325, 329, 330, 331, 332, 334, 335, 336, 337, 340, 341, 343, 344, 345, 349, 350, 351, 355, 356, 357, 358, 359, 360, 362, 363, 364, 365, 366, 370, 371, 372, 373, 375, 377, 378, 379, 380, 381, 384, 385, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 400, 402, 999, 1000]|
|:large_blue_circle:|com/starrocks/catalog/Partition.java|19|53|35.85%|[237, 238, 239, 240, 242, 243, 244, 245, 246, 247, 250, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 363, 529, 539, 542, 543, 544, 546, 547, 549]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndexMeta.java|8|22|36.36%|[210, 211, 213, 214, 216, 217, 223, 227, 228, 235, 236, 239, 243, 244]|
|:large_blue_circle:|com/starrocks/transaction/OlapTableTxnStateListener.java|1|2|50.00%|[155]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|1|2|50.00%|[355]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|4|7|57.14%|[1496, 1591, 1592]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndex.java|7|12|58.33%|[216, 220, 221, 228, 229]|
|:large_blue_circle:|com/starrocks/load/streamload/StreamLoadTask.java|2|3|66.67%|[873]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|18|26|69.23%|[1396, 1397, 1836, 1837, 1838, 1839, 1841, 1842]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|10|12|83.33%|[269, 270]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/load/DeleteMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/OlapTableTxnLogApplier.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/TransactionState.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateMaterializedViewStmt.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 162 / 477 (33.96%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Table.java|0|1|00.00%|[230]|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|7|00.00%|[509, 525, 527, 528, 531, 534, 1124]|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|0|1|00.00%|[117]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[2897, 2898]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|10|00.00%|[244, 245, 246, 247, 248, 249, 250, 251, 1164, 1165]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|4|00.00%|[253, 254, 255, 256]|
|:large_blue_circle:|com/starrocks/persist/CreateMaterializedIndexMetaInfo.java|0|34|00.00%|[30, 37, 39, 42, 43, 44, 45, 46, 47, 50, 54, 58, 62, 66, 67, 68, 69, 70, 73, 74, 75, 76, 77, 80, 81, 82, 87, 92, 93, 95, 96, 98, 99, 100]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|16|06.25%|[1624, 2072, 2073, 2074, 2076, 2077, 2078, 2080, 2081, 2083, 2084, 2085, 2086, 2088, 2091]|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|19|121|15.70%|[226, 252, 253, 254, 255, 256, 258, 259, 260, 262, 264, 265, 269, 270, 271, 272, 273, 274, 275, 276, 279, 280, 281, 282, 283, 284, 285, 286, 288, 289, 290, 295, 296, 297, 298, 299, 302, 303, 308, 309, 311, 312, 315, 316, 317, 318, 319, 321, 323, 324, 326, 330, 331, 332, 333, 335, 336, 337, 338, 341, 342, 344, 345, 346, 350, 351, 352, 356, 357, 358, 359, 360, 361, 363, 364, 365, 366, 367, 371, 372, 373, 374, 376, 379, 380, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 397, 994, 995]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|5|18|27.78%|[472, 475, 564, 566, 568, 570, 587, 589, 1092, 1093, 1094, 1095, 1097]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndexMeta.java|8|22|36.36%|[210, 211, 213, 214, 216, 217, 223, 227, 228, 235, 236, 239, 243, 244]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|56|154|36.36%|[228, 229, 230, 231, 232, 235, 239, 243, 266, 267, 268, 269, 270, 271, 350, 351, 352, 353, 354, 355, 356, 358, 360, 361, 363, 364, 366, 367, 368, 369, 371, 372, 373, 374, 375, 376, 377, 380, 381, 382, 383, 384, 387, 388, 389, 390, 391, 392, 394, 395, 396, 397, 398, 399, 401, 402, 403, 405, 406, 407, 408, 570, 571, 573, 574, 575, 576, 577, 580, 581, 582, 583, 584, 585, 588, 590, 631, 632, 633, 634, 635, 637, 638, 639, 640, 642, 643, 644, 645, 669, 670, 671, 672, 673, 680, 682, 683, 712]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java|2|5|40.00%|[201, 202, 698]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|1|2|50.00%|[355]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|4|7|57.14%|[1497, 1592, 1593]|
|:large_blue_circle:|com/starrocks/alter/AlterJobMgr.java|2|3|66.67%|[232]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|10|12|83.33%|[269, 270]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|22|25|88.00%|[568, 1396, 1397]|
|:large_blue_circle:|com/starrocks/transaction/GlobalTransactionMgr.java|12|13|92.31%|[323]|
|:large_blue_circle:|com/starrocks/catalog/Database.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/MetadataMgr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/OlapTableTxnLogApplier.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/TransactionState.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateMaterializedViewStmt.java|3|3|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 1]
```diff
@@ -568,6 +568,7 @@ createMaterializedViewStatement
     ('(' columnNameWithComment (',' columnNameWithComment)* ')')?
     comment?
     materializedViewDesc*
+    (TO targetTable=qualifiedName)?
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:571]: Just one "TO" I think is not intuitive enough to understand the specific meaning of the grammar

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:580]: Do you have any better idea? this is follow CK's syntax(https://clickhouse.com/docs/en/sql-reference/statements/create/view#materialized-view)

File: be/src/exec/tablet_info.cpp [hunk 2]
```diff
@@ -134,6 +135,8 @@ OlapTablePartitionParam::OlapTablePartitionParam(std::shared_ptr<OlapTableSchema
 OlapTablePartitionParam::~OlapTablePartitionParam() = default;
 
 Status OlapTablePartitionParam::init(RuntimeState* state) {
+    VLOG(2) << "OlapTablePartitionParam partition init:\n" << apache::thrift::ThriftDebugString(_t_param);
```
Comment (chaoyli) [inline be/src/exec/tablet_info.cpp:138]: 1. Which expressions are supported in the single table sync materialized view?
2. Does the single materialized view use insert overwrite or not?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 262 / 321 (81.62%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[2969, 2970]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|4|00.00%|[263, 264, 265, 266]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|9|22.22%|[257, 258, 259, 260, 261, 262, 265]|
|:large_blue_circle:|com/starrocks/transaction/GlobalTransactionMgr.java|4|8|50.00%|[326, 334, 335, 336]|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|85|107|79.44%|[257, 283, 285, 286, 287, 288, 289, 290, 291, 292, 293, 333, 401, 404, 405, 406, 1000, 1023, 1024, 1027, 1028, 1030]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|76|94|80.85%|[242, 263, 377, 390, 397, 398, 399, 408, 409, 410, 411, 416, 418, 422, 473, 594, 618, 622]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateMaterializedViewStmt.java|11|12|91.67%|[392]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|18|19|94.74%|[2067]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndexMeta.java|15|15|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Table.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MVUtils.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|28|28|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/FileScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/MVColumnItem.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/StreamLoadScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/TransactionState.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 219 / 439 (49.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Table.java|0|1|00.00%|[239]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MVUtils.java|0|3|00.00%|[91, 92, 94]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[2979, 2980]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|4|00.00%|[263, 264, 265, 266]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|18|05.56%|[1613, 2073, 2074, 2075, 2076, 2078, 2079, 2080, 2083, 2084, 2085, 2086, 2087, 2088, 2089, 2091, 2093]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|2|10|20.00%|[257, 258, 259, 260, 261, 262, 263, 264]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|50|154|32.47%|[227, 228, 229, 230, 231, 234, 238, 242, 264, 269, 270, 271, 272, 273, 274, 356, 357, 358, 359, 360, 361, 362, 364, 366, 367, 369, 370, 372, 373, 374, 375, 377, 378, 379, 380, 381, 382, 383, 386, 387, 388, 389, 390, 393, 394, 395, 396, 397, 398, 400, 401, 402, 403, 404, 405, 407, 408, 409, 411, 412, 413, 414, 431, 432, 577, 578, 580, 581, 582, 583, 584, 585, 586, 587, 588, 589, 590, 591, 592, 593, 595, 597, 638, 639, 640, 641, 642, 644, 645, 646, 647, 649, 650, 651, 652, 716, 717, 718, 719, 720, 727, 729, 730, 745]|
|:large_blue_circle:|com/starrocks/persist/CreateMaterializedIndexMetaInfo.java|12|34|35.29%|[37, 39, 50, 54, 58, 62, 73, 74, 75, 76, 77, 80, 81, 82, 87, 92, 93, 95, 96, 98, 99, 100]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java|2|5|40.00%|[202, 203, 712]|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|87|132|65.91%|[254, 257, 258, 263, 265, 272, 294, 295, 313, 327, 328, 331, 332, 337, 338, 340, 341, 344, 345, 346, 347, 348, 350, 352, 353, 355, 359, 360, 361, 362, 364, 365, 366, 367, 375, 381, 385, 386, 387, 412, 421, 422, 1000, 1048, 1049]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedIndexMeta.java|15|22|68.18%|[245, 246, 248, 249, 251, 252, 274]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateMaterializedViewStmt.java|7|8|87.50%|[390]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|21|23|91.30%|[1576, 1577]|
|:large_blue_circle:|com/starrocks/transaction/GlobalTransactionMgr.java|12|13|92.31%|[321]|
|:large_blue_circle:|com/starrocks/planner/FileScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/MVColumnItem.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/StreamLoadScanNode.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/transaction/TransactionState.java|5|5|100.00%|[]|

**Pull request closed without merging**